### PR TITLE
feat(Slice4): move and broadcast by websocket Squashed commit of the …

### PR DESCRIPTION
**Pull request by ra121e**
…following:

commit 72d7d02d4430a39fc9171dab131a7ed6e4d3ce82
Author: root <sushi.honda@gmail.com>
Date:   Wed Apr 29 11:00:15 2026 +0800

    fix(Slice4): format error unused timeout variable

    Failed to type check.
    ./app/api/matches/[id]/moves/route.ts:173:13
    Type error: 'timeoutMs' is declared but its value is never read.
      171 |     // that was already persisted.
        172 |     try {
    	> 173 |       const timeoutMs = Number(process.env["REALTIME_PUBLISH_TIMEOUT_MS"] ?? 2000);
    	      |             ^
    		    174 |       await publishGameUpdate(gameUpdate).catch((e) => Promise.reject(e));
    			  175 |       // Note: we intentionally await here to keep ordering guarantees for
    			    176 |       // connected clients; if you prefer fire-and-forget, call without await.
    				Next.js build worker exited with code: 1 and signal: null
    				error: script "build" exited with code 1
    				Error: Process completed with exit code 1.

commit 3b5458e07b80ed196599813fd6f03fb91604843e
Author: root <sushi.honda@gmail.com>
Date:   Wed Apr 29 10:51:08 2026 +0800

    fix(Slice4): Avoid treating committed moves as failed when realtime broadcast fails

      - Persist moves inside the Prisma transaction and move realtime publish outside of it. Add a short publish timeout (AbortController; REALTIME_PUBLISH_TIMEOUT_MS, default 2000ms) and catch/log publish errors so broadcast failures do not cause a 500 for an already-committed move. Add a TODO note to implement an outbox/retry worker for reliable delivery.

commit 7de16f3f00e8ca1b0156a160a8010aee5bd0f8f0
Author: root <sushi.honda@gmail.com>
Date:   Mon Apr 27 12:29:17 2026 +0800

    feat(slice4:WIP): In Frontend page.ts, add Move card section

commit 3e850a78112b7471d430a12d1945a881527a2d10
Author: root <sushi.honda@gmail.com>
Date:   Mon Apr 27 11:47:40 2026 +0800

    feat(Slice4:WIP): In Frontend component, add move form

commit f102e2ba929f5024ee63653600640fc43d5e0d2d
Author: root <sushi.honda@gmail.com>
Date:   Mon Apr 27 10:05:47 2026 +0800

    feat(Slice4:WIP): In frontend, add game:update event handler

    Extend `useSocketGame` to receive and store `game:update` payloads from the realtime server.

    - add `lastUpdate` state for the latest game payload
    - reset stored update when the match session changes
    - subscribe to `game:update` alongside existing subscription status handling

commit 5920f63892602a56cbf449ceff075f41002d36e6
Author: root <sushi.honda@gmail.com>
Date:   Mon Apr 27 09:39:48 2026 +0800

    feat(Slice4:WIP): add routing in Bun server

commit 21bdbcf0fba4ce07abe36a64c7689b9c2e8c5fb2
Author: athonda <athonda@vbox.42singapore.sg>
Date:   Sat Apr 25 15:34:26 2026 +0800

    feat(Slice4:WIP): broadcast by realtime server

commit 5339cefaee21943ea651a7d481919c8b61944828
Author: athonda <athonda@vbox.42singapore.sg>
Date:   Sat Apr 25 13:43:18 2026 +0800

    fix(slice4:WIP): correct move payload validation and lastMove participantId key

    ./app/api/matches/[id]/moves/route.ts:114:11
    Type error: Type 'number | undefined' is not assignable to type 'number'.
      Type 'undefined' is not assignable to type 'number'.

      112 |           participantId: participant.id,
      113 |           moveNumber: moveCount + 1,
    > 114 |           x: position.x,
          |           ^
      115 |           y: position.y,
      116 |           requestId,
      117 |           baseVersion,
    Next.js build worker exited with code: 1 and signal: null
    error: script "build" exited with code 1
    Error: Process completed with exit code 1.

commit e01710fc1f83fe9c3aabc8b701dd2a26dc3ac194
Author: root <sushi.honda@gmail.com>
Date:   Sat Apr 25 07:18:17 2026 +0800

    feat(slice4:WIP): add minimal move submission route

    Add the initial `POST /api/matches/:id/moves` Route Handler for the
    Slice 4 vertical slice.

      - accept move submission requests with `participantId`, `position`, `requestId`, and `baseVersion`
      - persist `MatchMove` and advance `Match.stateVersion` / `nextTurnSeat` - return basic accepted/error responses as the first server-side move flow

commit 9e4a661788177435b8aa8b21925d173ab5c02c74
Author: root <sushi.honda@gmail.com>
Date:   Fri Apr 24 22:08:38 2026 +0800

    doc: revise dev-plan-phase0.md and rename

commit b69c9dcfeefb74fdf9428649cfd01ab24fae0ee0
Author: root <sushi.honda@gmail.com>
Date:   Fri Apr 24 21:02:41 2026 +0800

    feat(slice4:WIP): add common type and interfaces

      - define request/ack payloads for `match:subscribe` and `match:subscribed`
      - add `GameUpdatePayload` and related interfaces for `participants`, `lastMove`, match status, seats, and board state - standardize on `participantId` so subscription and update events use the same identifier across the app

### CODE REVIEW

File: app/api/matches/[id]/moves/route.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+  return doFetch(timeoutMs);
+}
+
+export async function POST(request: Request, { params }: { params: Promise<{ id: string }> }) {
+  try {
+    const { id: matchId } = await params;
+    const body = (await request.json()) as SubmitMoveRequest;
+
+    if (!body.participantId || !isValidPosition(body.position)) {
```
Comment (chrisgohmz) [inline app/api/matches/[id]/moves/route.ts:95]: This validates payload shape, but the route still persists and broadcasts moves without enforcing the core match invariants: board bounds, current turn, stale `baseVersion`, and occupied cells. Because this is the command endpoint that advances `stateVersion`, those checks need to happen inside the transaction before `matchMove.create`; otherwise invalid moves become the official DB and realtime state.

File: shared/match-events.ts [hunk 2]
```diff
@@ -6,3 +6,36 @@ export type MatchSubscribePayload = {
...
+  status: MatchStatus;
+  // visibility: "PUBLIC" | "PRIVATE";
+  // ruleType: "GOMOKU" | "RENJU";
+  // boardSize: number;
+  stateVersion: number;
+  nextTurnSeat: Seat | null;
+  // winningSeat: Seat | null;
+  // endReason: string | null;
+  // participants: ParticipantSummary[];
+  lastMove: LastMove | null;
```
Comment (chrisgohmz) [inline shared/match-events.ts:39]: The shared contract says `lastMove` can be `null`, but `realtime/server.ts` rejects payloads without `lastMove`. Either make the internal publish payload a narrower “move update” type where `lastMove` is required, or let the realtime validator accept the same nullable shape.

File: realtime/server.ts [hunk 3]
```diff
@@ -6,7 +6,9 @@ import { Server as Engine } from "@socket.io/bun-engine";
 import { config } from "dotenv";
 import { Server } from "socket.io";
 
+import type { GameUpdatePayload } from "../app/shared/match-events";
```
Comment (chrisgohmz) [inline realtime/server.ts]: This makes the standalone realtime service depend on an app-owned directory. Since these event contracts are shared across app, client, and realtime, they should live in a root `shared/` module/package so neither process depends on the other's internal layout.

File: app/components/proto/proto-client.tsx [hunk 4]
```diff
@@ -192,7 +198,47 @@ export function ProtoClient() {
           </ul>
         </article>
         <article className="card">
-          <p>{t("descriptionStatus", { status })}</p>
+          <p>discription status: {status}</p>
```
Comment (chrisgohmz) [inline app/components/proto/proto-client.tsx]: This regresses the existing localization path and introduces a visible typo (`discription`). The previous `t("descriptionStatus", { status })` message already exists in en/ja/zh, so the new move/status copy should stay in `next-intl` too.

**Review by chrisgohmz — CHANGES_REQUESTED** [review]
# Brooks-Lint Review

**Mode:** PR Review
**Scope:** PR #75 vs `main`; reviewed production changes in API, realtime, shared event types, and proto UI. Large doc rename noted but not treated as production risk.
**Health Score:** 69/100

This slice builds the intended move/broadcast path, but the server currently accepts impossible game states, so I would not merge it as-is.

## Findings

### Critical

**Domain Model Distortion — Move route does not own core game invariants**
Symptom: `app/api/matches/[id]/moves/route.ts` validates only integer `x/y` plus participant existence; it never checks board bounds, `participant.seat === match.nextTurnSeat`, stale `baseVersion`, or occupied cell before persisting and broadcasting.
Source: *Domain-Driven Design* — Aggregate Roots / invariant ownership
Consequence: A player can submit repeated turns or off-board moves, and the DB plus `game:update` stream become the official source of an invalid match.
Remedy: Enforce move invariants inside the transaction before `matchMove.create`: bounds from `match.boardSize`, player turn, optional `baseVersion`, and occupied-cell checks with specific REST errors.

### Warning

**Knowledge Duplication — Event contract says `lastMove` is nullable, runtime rejects it**
Symptom: `GameUpdatePayload.lastMove` allows `null`, but realtime returns `invalid_payload` when `lastMove` is absent.
Source: *The Pragmatic Programmer* — DRY, duplicated knowledge
Consequence: Future status-only updates can type-check in the app and still fail at the realtime boundary.
Remedy: Either narrow the internal publish payload to “move update with lastMove” or align the realtime validator with `GameUpdatePayload`.

**Dependency Disorder — Realtime depends on an app-owned type module**
Symptom: `realtime/server.ts` imports shared event contracts from `../app/shared/match-events`.
Source: *Clean Architecture* — Dependency direction / conceptual boundary
Consequence: The standalone realtime service now depends on the Next app directory layout; moving app internals can break the socket server.
Remedy: Move event contracts to a root `shared/` module or package and import that from both `app` and `realtime`.

**Coverage Illusion — New public move/broadcast behavior has no tests**
Symptom: PR #75 adds `POST /api/matches/[id]/moves`, realtime publishing, and client submission, but no matching test files were added.
Source: *Working Effectively with Legacy Code* — Characterization tests before behavior changes
Consequence: The missing invariant cases above can regress silently because neither route behavior nor event validation is covered.
Remedy: Add focused tests for accepted move, wrong turn, out-of-bounds, occupied cell, duplicate `requestId`, and realtime invalid-payload handling.

### Suggestion

**Knowledge Duplication — Proto UI bypasses existing translations**
Symptom: `proto-client.tsx` replaces `t("descriptionStatus", { status })` with hard-coded English text, including the typo `discription`.
Source: *Domain-Driven Design* — Ubiquitous Language drift
Consequence: The `/ja/proto` and `/zh/proto` pages lose localized copy in the new move panel.
Remedy: Keep using `next-intl` messages for the new move/status strings.

## Verification

- `bun run lint` passed.
- `bun run build` passed.
- Browser verification was not completed because Python Playwright is not installed in the workspace.

File: app/api/matches/[id]/moves/route.ts (hunk 1 above)
Comment (chrisgohmz) [reply app/api/matches/[id]/moves/route.ts:95]: Addressed in `a0b81ab`: move validation now runs inside the transaction before `matchMove.create`, covering board bounds, current turn, stale `baseVersion`, duplicate `requestId`, and occupied cells. I also added focused Bun tests for the rule outcomes.

File: shared/match-events.ts (hunk 2 above)
Comment (chrisgohmz) [reply shared/match-events.ts:39]: Addressed in `a0b81ab`: realtime now validates against a shared payload guard that accepts `lastMove: null` when the contract allows it, with tests covering both move updates and status-only updates.

File: realtime/server.ts (hunk 3 above)
Comment (chrisgohmz) [reply realtime/server.ts]: Addressed in `a0b81ab`: moved the event contracts to root `shared/` and updated both app and realtime imports so the realtime service no longer depends on `app/shared`.

File: app/components/proto/proto-client.tsx (hunk 4 above)
Comment (chrisgohmz) [reply app/components/proto/proto-client.tsx]: Addressed in `a0b81ab`: restored `next-intl` usage in the proto panel and added localized en/ja/zh message keys for the new move/status text.

**Review by chrisgohmz — APPROVED** [review]
